### PR TITLE
File data logging

### PR DESCRIPTION
**Pull request by regit**
Make sure these boxes are checked accordingly before submitting your Pull Request -- thank you.

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [x] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [x] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to tickets:
- https://redmine.openinfosecfoundation.org/issues/7346
- https://redmine.openinfosecfoundation.org/issues/7347

Describe changes:
- Don't log sha256 on incomplete files
- Optional logging of file data in alerts

### Provide values to any of the below to override the defaults.

SV_REPO=https://github.com/regit/suricata-verify/
SV_BRANCH=issue-7346

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on MULTI_SMB_FILES_sha256.

ERROR: QA failed on SURI_TLPW1_files_sha256.
  

Pipeline 23189

**Comment by regit** [comment]
I've just added a missing sanity check as review was not yet started.

**Comment by regit** [comment]
> ERROR:
> 
> ERROR: QA failed on MULTI_SMB_FILES_sha256.
> 
> ERROR: QA failed on SURI_TLPW1_files_sha256.
> 
> Pipeline 23189

This should be link to sha256 computed on not fully captured file.

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on MULTI_SMB_FILES_sha256.

ERROR: QA failed on SURI_TLPW1_files_sha256.
  

Pipeline 23190

File: src/output-json.c [hunk 1]
```diff
@@ -159,10 +164,6 @@ void EveFileInfo(JsonBuilder *jb, const File *ff, const uint64_t tx_id, const ui
             break;
     }
 
-    if (ff->flags & FILE_SHA256) {
-        jb_set_hex(jb, "sha256", (uint8_t *)ff->sha256, (uint32_t)sizeof(ff->sha256));
-    }
```
Comment (jasonish) [inline src/output-json.c:164]: This was always logged as even truncated, these files are written into the file store with this sha256 as its filename. So lets you associate this record with the logged file. Probably should have been a comment here.

Comment (regit) [reply src/output-json.c:164]: ok, I should have thought about that. I'm going to remove it.

Comment (regit) [reply src/output-json.c:164]: I'm wondering if if could help to do something like:
```diff
diff --git a/src/output-json.c b/src/output-json.c
index 4815665f1b..f3b74bc58c 100644
--- a/src/output-json.c
+++ b/src/output-json.c
@@ -149,6 +149,9 @@ void EveFileInfo(JsonBuilder *jb, const File *ff, const uint64_t tx_id, const ui
             if (ff->flags & FILE_SHA1) {
                 jb_set_hex(jb, "sha1", (uint8_t *)ff->sha1, (uint32_t)sizeof(ff->sha1));
             }
+            if (ff->flags & FILE_SHA256) {
+                jb_set_hex(jb, "sha256", (uint8_t *)ff->sha256, (uint32_t)sizeof(ff->sha256));
+            }
             break;
         case FILE_STATE_TRUNCATED:
             JB_SET_STRING(jb, "state", "TRUNCATED");
@@ -161,13 +164,12 @@ void EveFileInfo(JsonBuilder *jb, const File *ff, const uint64_t tx_id, const ui
             break;
     }
 
-    if (ff->flags & FILE_SHA256) {
-        jb_set_hex(jb, "sha256", (uint8_t *)ff->sha256, (uint32_t)sizeof(ff->sha256));
-    }
-
     if (flags & FILE_STORED) {
         JB_SET_TRUE(jb, "stored");
         jb_set_uint(jb, "file_id", ff->file_store_id);
+        if ((ff->state != FILE_STATE_CLOSED) && (ff->flags & FILE_SHA256)) {
+            jb_set_hex(jb, "sha256", (uint8_t *)ff->sha256, (uint32_t)sizeof(ff->sha256));
+        }
     } else {
         JB_SET_FALSE(jb, "stored");
         if (flags & FILE_STORE) {
```
where we log only if file is stored or complete ?

Comment (regit) [reply src/output-json.c:164]: My proposal is stupid: incomplete files are interesting and need a sha256 to find them. I'm going to close this PR soon and do an updated one without the patch.

**Pull request closed without merging**